### PR TITLE
Adding pytorch code for 05-cnn

### PR DESCRIPTION
**Pull request by MysteryVaibhav**
Numbers for both the versions match.

**dynet**
```
iter 0: train loss/sent=1.5063, acc=0.3188, time=28.13s
iter 0: test acc=0.3416
iter 1: train loss/sent=1.2567, acc=0.4671, time=32.97s
iter 1: test acc=0.3842
iter 2: train loss/sent=0.9572, acc=0.6323, time=32.45s
iter 2: test acc=0.3706
iter 3: train loss/sent=0.6188, acc=0.7869, time=29.11s
iter 3: test acc=0.3887
iter 4: train loss/sent=0.3318, acc=0.8960, time=29.59s
iter 4: test acc=0.3878
iter 5: train loss/sent=0.1578, acc=0.9544, time=30.80s
iter 5: test acc=0.3738
iter 6: train loss/sent=0.0697, acc=0.9812, time=29.31s
iter 6: test acc=0.3810
iter 7: train loss/sent=0.0289, acc=0.9927, time=30.40s
iter 7: test acc=0.3751
iter 8: train loss/sent=0.0142, acc=0.9971, time=29.38s
iter 8: test acc=0.3846
iter 9: train loss/sent=0.0080, acc=0.9980, time=33.48s
iter 9: test acc=0.3765
```
**pytorch**
```
iter 0: train loss/sent=1.4739, acc=0.3415, time=27.00s
iter 0: test acc=0.4077
iter 1: train loss/sent=1.0367, acc=0.5846, time=27.34s
iter 1: test acc=0.3995
iter 2: train loss/sent=0.4667, acc=0.8471, time=28.26s
iter 2: test acc=0.3787
iter 3: train loss/sent=0.1347, acc=0.9627, time=27.34s
iter 3: test acc=0.3923
iter 4: train loss/sent=0.0397, acc=0.9886, time=28.04s
iter 4: test acc=0.3860
iter 5: train loss/sent=0.0168, acc=0.9951, time=27.58s
iter 5: test acc=0.3443
iter 6: train loss/sent=0.0067, acc=0.9978, time=28.90s
iter 6: test acc=0.3570
iter 7: train loss/sent=0.0034, acc=0.9994, time=26.13s
iter 7: test acc=0.3801
iter 8: train loss/sent=0.0014, acc=0.9998, time=28.74s
iter 8: test acc=0.3570
iter 9: train loss/sent=0.0022, acc=0.9996, time=28.36s
iter 9: test acc=0.3769
```

### CODE REVIEW

**Comment by MysteryVaibhav** [comment]
@danishpruthi: Please review the code whenever you get time. Only difference which I can see is nn.Conv1d() in torch vs dy.conv2d_bias() in dynet. I believe that they will have exactly the same number of parameters.

File: 05-cnn-pytorch/cnn-activation.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+        self.relu = torch.nn.ReLU()
+        self.projection_layer = torch.nn.Linear(in_features=num_filters, out_features=ntags, bias=True)
+        # Initializing the projection layer
+        torch.nn.init.xavier_uniform_(self.projection_layer.weight)
+
+    def forward(self, words, return_activations=False):
+        emb = self.embedding(words)                 # nwords x emb_size
+        emb = emb.unsqueeze(0).permute(0, 2, 1)     # 1 x emb_size x nwords
+        h = self.conv_1d(emb)                       # 1 x num_filters x nwords
+        activations = h.squeeze().max(dim=1)[1]     # argmax along length of the sentence
```
Comment (muk343) [inline 05-cnn-pytorch/cnn-activation.py]: I think it would be better to have something like this: 
activations = h.squeeze(0).max(dim=1)[1]
That is, specifically saying which particular dimension to squeeze. Else, for the cases where nwords = window_size, h would have a dimension of 1xnum_filterx1. Then, pytorch would squeeze both the first and last dimension and finally, we would be left with a tensor with only one dimension. And there selection "[1]" would fail.

Comment (MysteryVaibhav) [reply 05-cnn-pytorch/cnn-activation.py]: Good suggestion to make the squeeze dimension explicit.

File: 05-cnn-pytorch/cnn-class.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        # Initializing the projection layer
+        torch.nn.init.xavier_uniform_(self.projection_layer.weight)
+
+    def forward(self, words):
+        emb = self.embedding(words)                 # nwords x emb_size
+        emb = emb.unsqueeze(0).permute(0, 2, 1)     # 1 x emb_size x nwords
+        h = self.conv_1d(emb)                       # 1 x num_filters x nwords
+        # Do max pooling
+        h = h.max(dim=2)[0]                         # 1 x num_filters
+        h = self.relu(h)
```
Comment (danishpruthi) [inline 05-cnn-pytorch/cnn-class.py:29]: I am not sure if the `relu` activation here has a correspondence to dynet's `cnn-class.py`. I couldn't' find it. Is there any particular reason for clamping the negative values to zero?

File: 05-cnn-pytorch/cnn-activation.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+        words_tensor = torch.tensor(wids).type(type)
+        scores = model(words_tensor)
+        predict = scores[0].argmax().item()
+        if predict == tag:
+            test_correct += 1
+    print("iter %r: test acc=%.4f" % (ITER, test_correct/len(dev)))
+
+
+for words, wids, tag in dev:
+    calc_predict_and_activations(wids, tag, words)
```
Comment (danishpruthi) [inline 05-cnn-pytorch/cnn-activation.py]: maybe add an `input()` prompt below this, so that people can press any key, and view more examples (like in the original version)

File: 05-cnn-pytorch/cnn-activation.py (hunk 1 above)
Comment (danishpruthi) [reply 05-cnn-pytorch/cnn-activation.py]: @MysteryVaibhav, just FYI, the code crashes for an example in the dataset because of this issue. So maybe, explicitly add the dimension.

File: 05-cnn-pytorch/cnn-activation.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        torch.nn.init.xavier_uniform_(self.projection_layer.weight)
+
+    def forward(self, words, return_activations=False):
+        emb = self.embedding(words)                 # nwords x emb_size
+        emb = emb.unsqueeze(0).permute(0, 2, 1)     # 1 x emb_size x nwords
+        h = self.conv_1d(emb)                       # 1 x num_filters x nwords
+        activations = h.squeeze().max(dim=1)[1]     # argmax along length of the sentence
+        # Do max pooling
+        h = h.max(dim=2)[0]                         # 1 x num_filters
+        h = self.relu(h)
```
Comment (danishpruthi) [inline 05-cnn-pytorch/cnn-activation.py:31]: Similar to above, I am not sure if we need a `relu` here.

**Review by danishpruthi — CHANGES_REQUESTED** [review]
Hi @MysteryVaibhav, this mostly looks good 👍 

I have a few minor comments, please address those, and then we can get this one in. Sorry for taking long to have a look.

File: 05-cnn-pytorch/cnn-class.py (hunk 2 above)
Comment (MysteryVaibhav) [reply 05-cnn-pytorch/cnn-class.py:29]: @danishpruthi It's corresponding to dy.rectify(pool_out)

File: 05-cnn-pytorch/cnn-activation.py (hunk 3 above)
Comment (MysteryVaibhav) [reply 05-cnn-pytorch/cnn-activation.py]: Thanks for this suggestion. Added.

File: 05-cnn-pytorch/cnn-activation.py (hunk 4 above)
Comment (MysteryVaibhav) [reply 05-cnn-pytorch/cnn-activation.py:31]: Resolved this.

**Comment by MysteryVaibhav** [comment]
@danishpruthi I have made the suggested changes. Please have a look again.

**Comment by danishpruthi** [comment]
This looks great @MysteryVaibhav . 

@neubig could you please merge this.
